src/lib/dnssd/IncrementalResolve.cpp

```diff
@@ -0,0 +1,374 @@
+
+    if (!mSpecificResolutionData.Valid())
+    {
+        flags.Set(RequiredInformation::kSrvInitialization);
+    }
+    else
+    {
+        if (mCommonResolutionData.numIPs == 0)
+        {
+            flags.Set(RequiredInformation::kIpAddress);
+        }
+    }
+
+    return flags;
+}
+
+CHIP_ERROR IncrementalResolver::OnRecord(const ResourceData & data, BytesRange packetRange)
+{
+    if (!IsActive())
+    {
+        return CHIP_NO_ERROR; // nothing to parse
+    }
+
+    switch (data.GetType())
+    {
+    case QType::PTR:
+        return OnPtrRecord(data, packetRange);
+    case QType::TXT:
+        if (data.GetName() != mRecordName.Get())
+        {
+            ChipLogDetail(Discovery, "TXT record received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+        return OnTxtRecord(data, packetRange);
+    case QType::A: {
+        if (data.GetName() != mServerName.Get())
+        {
+            ChipLogDetail(Discovery, "IP address received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+
+        Inet::IPAddress addr;
+        if (!ParseARecord(data.GetData(), &addr))
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        return OnIpAddress(addr);
+    }
+    case QType::AAAA: {
+        if (data.GetName() != mServerName.Get())
+        {
+            ChipLogDetail(Discovery, "IP address received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+
+        Inet::IPAddress addr;
+        if (!ParseAAAARecord(data.GetData(), &addr))
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        return OnIpAddress(addr);
+    }
+    case QType::SRV: // SRV handled on creation, ignored for 'additional data'
+    default:
+        // Other types not interesting during parsing
+        return CHIP_NO_ERROR;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR IncrementalResolver::OnPtrRecord(const ResourceData & data, BytesRange packetRange)
+{
+    // Here we handle subtype expectations. Data is of the form:
+    // <subtype>._sub._mattrc._udp.local or
+    // <subtype>._sub._mattrd._udp.local
+    //
+    // If these hold, then we have to check if PTR points at the current record and
+    // if yes, the subtype matches and information can be extracted.
+
+    if (!IsActiveCommissionParse())
+    {
+        ChipLogDetail(Discovery, "PTR record for non-commission resolution.");
+        return CHIP_NO_ERROR;
+    }
+
+    if (!IsCommissionSubtype(data.GetName()))
+    {
+        ChipLogDetail(Discovery, "PTR record that is not a sub-type for commissioning.");
+        return CHIP_NO_ERROR;
+    }
+
+    SerializedQNameIterator qname;
+
+    if (!ParsePtrRecord(data.GetData(), packetRange, &qname))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (qname != mRecordName.Get())
+    {
+        ChipLogDetail(Discovery, "PTR record that is not pointing to the current commission server.");
```

Comment:
"current commission server" is something I am not sure I understand in this context